fiona/_shim1.pyx

```diff
@@ -37,15 +53,36 @@ cdef void* gdal_open_vector(const char *path_c, int mode, drivers):
         cogr_ds = OGROpen(path_c, mode, NULL)
     return cogr_ds
 
-cdef void* gdal_create(void* cogr_driver, const char *path_c) except *:
-    cdef void* cogr_ds
+
+cdef void* gdal_create(void* cogr_driver, const char *path_c, options) except *:
+    cdef void* cogr_ds = NULL
+    cdef char **opts = NULL
+
+    encoding = options.get('encoding', None)
+    if encoding:
+        val = encoding.encode('utf-8')
+        CPLSetThreadLocalConfigOption('SHAPE_ENCODING', val)
+    else:
+        CPLSetThreadLocalConfigOption('SHAPE_ENCODING', "")
+
+    for k, v in options.items():
+        k = k.upper().encode('utf-8')
+        if isinstance(v, bool):
+            v = ('ON' if v else 'OFF').encode('utf-8')
+        else:
+            v = str(v).encode('utf-8')
+        log.debug("Set option %r: %r", k, v)
+        opts = CSLAddNameValue(opts, <const char *>k, <const char *>v)
+
     try:
         cogr_ds = exc_wrap_pointer(
             OGR_Dr_CreateDataSource(
-                cogr_driver, path_c, NULL))
+                cogr_driver, path_c, opts))
+        return cogr_ds
     except Exception as exc:
         raise DriverIOError(str(exc))
-    return cogr_ds
+    finally:
+        CSLDestroy(opts)
```

Comment:
@snorfalorpagus this work benefits enormously from your rewrite! In maint-1.7 I had to sprinkle `CSLDestroy(opts)` throughout because there were so many places we could exit.